### DISCUSSION TITLE
`X-Up-Events` not emitted on redirect responses

### DISCUSSION THREAD

**Discussion by matteocontrini** [unpoly/unpoly#514] [Q&A] [answered]
Hello,

I have a POST request to an endpoint that returns a 303 redirect and also tries to emit an event with the `X-Up-Events` response header.

However it seems that events are not emitted by Unpoly when the response is a non-200 (?) one.

I wanted to check if this is the intended behavior, and if so what's the reasoning behind it and what I might be doing wrong, or if it's a bug.

Thanks!

**Comment by matteocontrini** [comment]
Thinking about it, XHR automatically follows redirects so obviously Unpoly gets to see only the final response.

Is that the issue? Are there workarounds, perhaps?

**Comment by triskweline** [comment, marked as answer]
> it seems that events are not emitted by Unpoly when the response is a non-200 (?) one

`X-Up-Events` is used for both 200 and non-200 responses.

> Thinking about it, XHR automatically follows redirects so obviously Unpoly gets to see only the final response.

I also think that's the issue.

> Are there workarounds, perhaps?

I can think of two workarounds:

### Not redirecting

If you're processing an Unpoly request (`X-Up-Version` is present) you could render the destination route without redirecting.

The destination URL could be sent in the `X-Up-Location` header.


### Preserving response headers over redirects

A more involved solution is implemented by [unpoly-rails](https://github.com/unpoly/unpoly-rails), the Rails bindings for Unpoly.

When it sees that the user is redirecting after setting `X-Up-Events` it [encodes the response headers as query parameters](https://github.com/unpoly/unpoly-rails/blob/85831300cc0a9e92894ca38802fab83eda4e7b9e/lib/unpoly/rails/change.rb#L402-L439).

So instead of this (which would lose the X-Up-Events header):

```http
Location: /redirect-destination
X-Up-Events: <JSON>
```

It will send this:

```http
Location: /redirect-destination?_up_events=<JSON>
X-Up-Location: /redirect-destination
```

Note that we're also setting `X-Up-Location` so the `_up_events` parameter does not appear in the user-visible URL.

When `unpoly-rails` sees the `_up_events` parameter on the destination URL, it extracts the JSON and re-sets the `X-Up-Events` response header.

**Reply by matteocontrini** [reply]
Thanks a lot for the detailed answer. Those are interesting workarounds but I'm not sure if they fit with the server framework I'm currently using (ASP.NET Core), in the sense that they'd require a bit of work to make them work.

Before diving into that, would you mind if I give a bit more context about my use case? I hope I won't hijack too much this discussion.

I have a drawer layer where the user can see the details of an item and make edits to it. When the user changes something, I'd like to have some parts of the root layer (beneath the drawer) updated, since the information has changed.

These are the options I've tried:

- the one we're dicussing here, i.e. firing an event with `X-Up-Events`, which someone would listen in order to trigger an `up.reload()` on the root layer (or part of it)
- `up-on-dismissed="up.reload(...)"`, with the disadvantage of always reloading even when no changes were made while the layer was open. I expect this drawer to be quite trafficked (many open/close operations), so I don't really like this solution
- trying to listen to the completion of Unpoly requests that happen within the layer (and that produce edits), and trigger `up.reload()` upon that

I couldn't implement the third option. I wish there was an `up-on-loaded` attribute but the closest I've found is listening to `up:request:loaded` on the layer, which brings in some complications with registering/unregistering event handlers manually.

May I ask if I missed something and what would be the most "Unpoly way" of solving this problem?

Thank you! 😃

**Reply by triskweline** [reply]
When there was a change in the drawer, I would *accept* the drawer, not dismiss it. Then I could use `up-on-accepted="up.reload(...)"`, but do nothing in `up-on-dismissed`. When the user closes the dialog by pressing the close icon, clicking into backdrop, etc. that counts as dismissal, not acceptance.

**Reply by matteocontrini** [reply]
I thought about that but it works only if the drawer is closed (accepted) as soon as the change is done. In my case I'd like the drawer to stay open after edit actions (think about multiple buttons to perform different actions, and the user might want to click more than one without reopening the drawer every time).

**Reply by matteocontrini** [reply]
Not sure how I missed it, but `up-on-loaded` [actually exists](https://unpoly.com/a-up-follow#up-on-loaded) 😅

So something like the following seems to do the trick!

```
up-on-loaded="up.reload('main', { layer: 'root' })
```
